### ISSUE TITLE
tool_choice not enforced locally: 'none' and specific-function modes silently treated as 'auto'

### ISSUE THREAD

**Issue by raullenchai** [raullenchai/vllm-mlx#445]
## Summary
Server accepts `tool_choice` per OpenAI spec but never enforces it locally — all modes silently behave as `"auto"`. Affects both harmony (gpt-oss) and hybrid (Qwen3.5/3.6) families on `/v1/chat/completions`. Issue #442 is a special case of this gap (covers `"required"` only); this issue covers the other two modes.

Surfaced on 2026-05-22 fresh-PyPI v0.6.65 onboarding sweep against `mlx-community/gpt-oss-20b-MXFP4-Q8`.

### `tool_choice: "none"`
Spec: server MUST NOT call any tool — model should produce a natural-language answer using only the prompt.
Observed: model still emits `tool_calls`, identical to `tool_choice: "auto"`. The user gets a tool call they explicitly disabled.

### `tool_choice: {"type":"function","function":{"name":"X"}}`
Spec: server MUST call only function `X`, even if the model would prefer something else.
Observed: server returns whichever tool the model picks. Forced `get_time` came back as `get_weather` because the model preferred it.

### `tool_choice: "required"` (already filed as #442 for the leak symptom)
Spec: server MUST emit ≥1 tool call.
Observed: model hits `max_tokens` cap mid-analysis without committing; `tool_calls=[]`, `finish_reason="length"`.

## Root cause

`vllm_mlx/routes/chat.py:499-500` is the only place `tool_choice` is referenced — and only for the cloud-router passthrough:
```python
if request.tool_choice is not None:
    cloud_kwargs["tool_choice"] = request.tool_choice
```

No local enforcement anywhere. The local inference path:
- Doesn't strip tools from the prompt when `tool_choice == "none"`.
- Doesn't bias logits toward the named function when `tool_choice == {function: {name: X}}`.
- Doesn't constrain decoding to require a tool-call commentary channel when `tool_choice == "required"`.

## Suggested fix sketch

Two layers:

1. **Prompt-level** (cheap, partial coverage):
   - `tool_choice: "none"` → strip the `tools` field before chat-template rendering so the model never sees them. Most reliable signal.
   - `tool_choice: {function: {name: X}}` → render only the named function in the prompt, hiding the rest.

2. **Decoder-level** (correct, expensive):
   - `tool_choice: "required"` / specific-function → engage `outlines-core` (PR #132's FSM constrained decoder, currently paused as draft) to constrain the next tokens to a valid harmony commentary channel.
   - Alternative: post-generation retry policy — detect `tool_calls=[]` with `tool_choice="required"` and re-prompt with a stronger system message.

Layer 1 alone closes `none` and specific-function for the common case; `required` likely needs layer 2 or a retry policy (this is what #442 already tracks).

## Repro

```bash
rapid-mlx serve mlx-community/gpt-oss-20b-MXFP4-Q8 --port 8502 \
  --enable-auto-tool-choice --tool-call-parser harmony &

# Case A: tool_choice=none ignored
curl -s http://127.0.0.1:8502/v1/chat/completions -H 'Content-Type: application/json' -d '{
  "model":"mlx-community/gpt-oss-20b-MXFP4-Q8",
  "messages":[{"role":"user","content":"what'\''s the weather in Tokyo?"}],
  "tools":[{"type":"function","function":{"name":"get_weather","parameters":{"type":"object","properties":{"city":{"type":"string"}}}}}],
  "tool_choice":"none"
}' | jq '.choices[0].message.tool_calls'
# Expected: null. Observed: [{"id":"...","function":{"name":"get_weather", ...}}]

# Case B: specific-function not honored
curl -s http://127.0.0.1:8502/v1/chat/completions -H 'Content-Type: application/json' -d '{
  "model":"mlx-community/gpt-oss-20b-MXFP4-Q8",
  "messages":[{"role":"user","content":"what'\''s the weather in Tokyo?"}],
  "tools":[
    {"type":"function","function":{"name":"get_weather","parameters":{"type":"object","properties":{"city":{"type":"string"}}}}},
    {"type":"function","function":{"name":"get_time","parameters":{"type":"object","properties":{"city":{"type":"string"}}}}}
  ],
  "tool_choice":{"type":"function","function":{"name":"get_time"}}
}' | jq '.choices[0].message.tool_calls[0].function.name'
# Expected: "get_time". Observed: "get_weather"
```

## Scope notes
- Affects ALL families on local inference — harmony was the test surface, but hybrid Qwen3.5/3.6 with hermes parser exhibits the same gap (no local `tool_choice` enforcement code anywhere).
- Cloud-router passthrough is unaffected (delegates enforcement to the backing provider).
- #442 stays scoped to the `required`-mode-with-length-cap leak; this issue covers `none` and specific-function.

**Comment by raullenchai** [comment]
PR #446 (merged 2026-05-22) lands the prompt-level half of this issue:
- `tool_choice: "none"` → `request.tools = None` before chat-template render; model can't see tools, can't call them.
- `tool_choice: {type:function, function:{name:X}}` → filter `request.tools` to just X. HTTP 400 if X not present.
- `tool_choice: "auto"` / `"required"` / unset → tools untouched (`"required"` enforcement still tracked under #442).

6 new tests under `tests/test_chat_route_tool_choice_enforcement.py` pin each branch via the live FastAPI router with a recording engine. pr_validate MERGE-SAFE (7/7 active gates including 3×3 stress matrix on qwen3.5-35B / qwen3.6-27B / gpt-oss-20b). CI 7/7 green.

Marking this issue mostly-done — the `"required"` enforcement portion that this issue captured by reference to #442 remains, but the two cleanly-fixable modes are now correct. If you'd prefer to close this and treat #442 as the remaining canonical tracker, I can close. Leaving open for now as the umbrella for any future regressions on these two modes.
